[UnifiedPDF] Data detector borders are unstable during zoom

aprotyas opened the pull request:
#### 1f4e1577f355f4a93eb0f3c0a34e08459356654c
<pre>
[UnifiedPDF] Data detector borders are unstable during zoom
<a href="https://bugs.webkit.org/show_bug.cgi?id=274942">https://bugs.webkit.org/show_bug.cgi?id=274942</a>
<a href="https://rdar.apple.com/128865849">rdar://128865849</a>

Reviewed by Simon Fraser.

Any active data detector overlay is constantly repainted during a
magnification gesture because we used to call into
didInvalidateHighlightOverlayRects whenever we were in the plugin&apos;s
setScaleFactor method. This resulted in the overlay looking unstable and
a little jumpy when pinching in/out.

To deal with this, we follow the precedent set by ImageOverlayController
and dismiss any active highlight during a magnification gesture. In
addition, we disallow the DD overlay controller to present DD overlay
rects in general during a pinch zoom.

We achieve this by:

1. Introducing the canShowDataDetectorHighlightOverlays method on the
   plugin, which is consulted before highlight presentation.

2. Providing WebCore::DataDetectorHighlight instances the ability to be
   dismissed immediately, rather than just being faded out.

* Source/WebCore/platform/mac/DataDetectorHighlight.h:
* Source/WebCore/platform/mac/DataDetectorHighlight.mm:
(WebCore::DataDetectorHighlight::dismissImmediately):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::hideActiveHighlightOverlay):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::canShowDataDetectorHighlightOverlays const):
(WebKit::UnifiedPDFPlugin::didBeginMagnificationGesture):
(WebKit::UnifiedPDFPlugin::setScaleFactor):

Canonical link: <a href="https://commits.webkit.org/279581@main">https://commits.webkit.org/279581@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bd05389d6e9401e0d119fc1afe258157c7c136f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53837 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33205 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4560 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56141 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40701 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4422 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43592 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2991 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55935 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31413 "Found 3 new test failures: fast/events/ios/key-events-comprehensive/key-events-meta.html, fast/forms/form-submission-crash-4.html, webgl/pending/conformance2/glsl3/empty-shader-with-output.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46571 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24732 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3882 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2715 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58710 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29007 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4132 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51003 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46714 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50338 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11731 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31135 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -321,6 +321,13 @@ static String mutationObserverNotificationString()
 
+void UnifiedPDFPlugin::hideActiveDataDetectorHighlightOverlay()
+{
+    auto lastKnownMousePositionInDocumentSpace = convertDown<FloatPoint>(CoordinateSpace::Plugin, CoordinateSpace::PDFDocumentLayout, lastKnownMousePositionInView());
```
Seems a bit weird to use the last known mouse position to hide an active highlight that we must already be tracking somewhere? If you're getting pageIndex just to satisfy my future discrete mode, there will be easier ways (by just asking about which row is visible).

aprotyas replied:
This was a little forward looking, yes, but I suppose I can fix that if needed when necessary.

aprotyas replied:
> hide an active highlight that we must already be tracking somewhere

That's right, the overlay controller already tracks the active highlight, so I'll refrain from doing this.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/5bd05389d6e9401e0d119fc1afe258157c7c136f

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1281,6 +1288,11 @@ static String mutationObserverNotificationString()
+
+#if ENABLE(UNIFIED_PDF_DATA_DETECTION)
+    hideActiveDataDetectorHighlightOverlay();
+    dataDetectorOverlayController().setCanPresentHighlightOverlays(false);
```
Rather than setBlah(false) and setBlah(true), can't dataDetectorOverlayController call back to a function on the plugin like `isOKToShowHightlight() const { return ! m_inMagnificationGesture; }` ?

aprotyas replied:
Good point, thanks. This reduces how much state is floating around.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/5bd05389d6e9401e0d119fc1afe258157c7c136f

aprotyas committed:
[UnifiedPDF] Data detector borders are unstable during zoom
https://bugs.webkit.org/show_bug.cgi?id=274942
rdar://128865849

Reviewed by Simon Fraser.

Any active data detector overlay is constantly repainted during a
magnification gesture because we used to call into
didInvalidateHighlightOverlayRects whenever we were in the plugin's
setScaleFactor method. This resulted in the overlay looking unstable and
a little jumpy when pinching in/out.

To deal with this, we follow the precedent set by ImageOverlayController
and dismiss any active highlight during a magnification gesture. In
addition, we disallow the DD overlay controller to present DD overlay
rects in general during a pinch zoom.

We achieve this by:

1. Introducing the canShowDataDetectorHighlightOverlays method on the
   plugin, which is consulted before highlight presentation.

2. Providing WebCore::DataDetectorHighlight instances the ability to be
   dismissed immediately, rather than just being faded out.

* Source/WebCore/platform/mac/DataDetectorHighlight.h:
* Source/WebCore/platform/mac/DataDetectorHighlight.mm:
(WebCore::DataDetectorHighlight::dismissImmediately):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::hideActiveHighlightOverlay):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::canShowDataDetectorHighlightOverlays const):
(WebKit::UnifiedPDFPlugin::didBeginMagnificationGesture):
(WebKit::UnifiedPDFPlugin::setScaleFactor):

Canonical link: https://commits.webkit.org/279581@main
